Build wxQT toolkit on AppVeyor

MaartenBent opened the pull request:
Work in progress.

catalinr commented on src/qt/region.cpp:
```diff
@@ -106,7 +106,7 @@ wxRegion::wxRegion(const wxBitmap& bmp, const wxColour& transp, int tolerance)
     }
 
-    unsigned char raw[bmp.GetWidth()*bmp.GetHeight()];
+    unsigned char* raw = new unsigned char(bmp.GetWidth()*bmp.GetHeight());
```
Did you by any chance mean `[]` here? Considering the `memset` below, did you test this change?
What was wrong with the original code? Was the mem chunk too big for the stack?

catalinr replied:
Its length was not variable because it didn't change later. The old size was not a *compile time constant* which was indeed a problem. But you do need the square brackets, otherwise you get a pointer to only one char, initialized to `bmp.GetWidth()*bmp.GetHeight()` value, don't you?

MaartenBent replied:
Yes, thank you for noticing. I have fixed it.

vadz replied:
The best would be to use `wxScopedArray<unsigned char> raw(bmp.GetWidth()*bmp.GetHeight())` here.

MaartenBent commented on include/wx/qt/evtloop.h:
```diff
@@ -10,7 +10,7 @@
 class QTimer;
 
-class WXDLLIMPEXP_BASE wxQtEventLoopBase : public wxEventLoopBase
+class WXDLLIMPEXP_CORE wxQtEventLoopBase : public wxEventLoopBase
```
I changed it to fix the shared build (see warnings and errors). There is still a `wxGUIEventLoop` in `core`. But `wxQtEventLoopBase` is used by both gui and console event loops.

```
../../../wxWidgets/src/qt/evtloop.cpp:66:1: warning: 'wxQtEventLoopBase::wxQtEventLoopBase()' redeclared without dllimport attribute: previous dllimport ignored [-Wattributes]
 wxQtEventLoopBase::wxQtEventLoopBase()
 ^~~~~~~~~~~~~~~~~
../../../wxWidgets/src/qt/evtloop.cpp:83:1: warning: 'virtual wxQtEventLoopBase::~wxQtEventLoopBase()' redeclared without dllimport attribute after being referenced with dll linkage
 wxQtEventLoopBase::~wxQtEventLoopBase()
 ^~~~~~~~~~~~~~~~~
../../../wxWidgets/src/qt/evtloop.cpp:88:6: warning: 'virtual void wxQtEventLoopBase::ScheduleExit(int)' redeclared without dllimport attribute: previous dllimport ignored [-Wattributes]
 void wxQtEventLoopBase::ScheduleExit(int rc)
      ^~~~~~~~~~~~~~~~~
../../../wxWidgets/src/qt/evtloop.cpp:95:5: warning: 'virtual int wxQtEventLoopBase::DoRun()' redeclared without dllimport attribute: previous dllimport ignored [-Wattributes]
 int wxQtEventLoopBase::DoRun()
     ^~~~~~~~~~~~~~~~~
../../../wxWidgets/src/qt/evtloop.cpp:114:6: warning: 'virtual bool wxQtEventLoopBase::Pending() const' redeclared without dllimport attribute: previous dllimport ignored [-Wattributes]
 bool wxQtEventLoopBase::Pending() const
      ^~~~~~~~~~~~~~~~~
../../../wxWidgets/src/qt/evtloop.cpp:119:6: warning: 'virtual bool wxQtEventLoopBase::Dispatch()' redeclared without dllimport attribute: previous dllimport ignored [-Wattributes]
 bool wxQtEventLoopBase::Dispatch()
      ^~~~~~~~~~~~~~~~~
../../../wxWidgets/src/qt/evtloop.cpp:126:5: warning: 'virtual int wxQtEventLoopBase::DispatchTimeout(long unsigned int)' redeclared without dllimport attribute: previous dllimport ignored [-Wattributes]
 int wxQtEventLoopBase::DispatchTimeout(unsigned long timeout)
     ^~~~~~~~~~~~~~~~~
../../../wxWidgets/src/qt/evtloop.cpp:133:6: warning: 'virtual void wxQtEventLoopBase::WakeUp()' redeclared without dllimport attribute: previous dllimport ignored [-Wattributes]
 void wxQtEventLoopBase::WakeUp()
      ^~~~~~~~~~~~~~~~~
../../../wxWidgets/src/qt/evtloop.cpp:140:6: warning: 'virtual void wxQtEventLoopBase::DoYieldFor(long int)' redeclared without dllimport attribute: previous dllimport ignored [-Wattributes]
 void wxQtEventLoopBase::DoYieldFor(long eventsToProcess)
      ^~~~~~~~~~~~~~~~~
../../../wxWidgets/src/qt/evtloop.cpp: In member function 'virtual int wxQtEventLoopBase::DoRun()':
../../../wxWidgets/src/qt/evtloop.cpp:111:12: warning: 'ret' may be used uninitialized in this function [-Wmaybe-uninitialized]
     return ret;
            ^~~
```
```
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/8.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: coredll_event.o:event.cpp:(.text$_ZN14wxGUIEventLoopD1Ev[_ZN14wxGUIEventLoopD1Ev]+0xd): undefined reference to `__imp__ZN17wxQtEventLoopBaseD2Ev'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/8.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: coredll_event.o:event.cpp:(.text$_ZN14wxGUIEventLoopD0Ev[_ZN14wxGUIEventLoopD0Ev]+0x14): undefined reference to `__imp__ZN17wxQtEventLoopBaseD2Ev'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/8.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: coredll_event.o:event.cpp:(.text$_ZN16wxModalEventLoopD1Ev[_ZN16wxModalEventLoopD1Ev]+0xd): undefined reference to `__imp__ZN17wxQtEventLoopBaseD2Ev'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/8.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: coredll_event.o:event.cpp:(.text$_ZN16wxModalEventLoopD0Ev[_ZN16wxModalEventLoopD0Ev]+0x14): undefined reference to `__imp__ZN17wxQtEventLoopBaseD2Ev'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/8.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: coredll_qt_apptraits.o:apptraits.cpp:(.text$_ZN11wxEventLoopD1Ev[_ZN11wxEventLoopD1Ev]+0xd): undefined reference to `__imp__ZN17wxQtEventLoopBaseD2Ev'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/8.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: coredll_qt_apptraits.o:apptraits.cpp:(.text$_ZN11wxEventLoopD0Ev[_ZN11wxEventLoopD0Ev]+0x14): more undefined references to `__imp__ZN17wxQtEventLoopBaseD2Ev' follow
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/8.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: coredll_qt_evtloop.o:evtloop.cpp:(.text+0x88): undefined reference to `__imp__ZTV17wxQtEventLoopBase'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/8.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: coredll_qt_evtloop.o:evtloop.cpp:(.text+0x34c): undefined reference to `__imp__ZTV17wxQtEventLoopBase'
collect2.exe: error: ld returned 1 exit status
```

MaartenBent replied:
I don't need the fix for the CMake build btw, only when using configure. I'll revert it and see what AppVeyor does.

vadz commented on include/wx/evtloop.h:
```diff
@@ -284,6 +284,10 @@ class WXDLLIMPEXP_BASE wxEventLoopManual : public wxEventLoopBase
     #include "wx/msw/evtloopconsole.h"
 #endif
 
+#if defined(__WXQT__)
```
Is it normal that we include _both_ `wx/unix/evtloop.h` or `wx/msw/evtloopconsole.h` _and_ `wx/qt/evtloop.h`?

Because either `__UNIX__` or `__WINDOWS__` is still going to be defined, even when `__WXQT__` also is, isn't it?

Besides, the console event loop is really supposed to be tookit-independent, so that you could use the same base library with either GTK or Qt-based core, for example. It would be nice if we could keep it like this.

MaartenBent replied:
It is probably not correct, but I tried the get a successful build by changing as little `WXQT`-code as possible.

Yes, in my case `__WINDOWS__ ` is still defined and the Windows `wxConsoleEventLoop` is compiled as well. That is why I had to rename the `WXQT` console event loop to `wxQtConsoleEventLoop`.

Following your paradigm, `WXQT` should not define a console event loop? And all `WXQT` code should be part of `core`?

vadz reviewed (commented):
Thanks for the fixes, please let me know when will you want to have this merged, but I don't really understand what's going on with the event loops here. Perhaps we need to discuss this on wx-dev?

MaartenBent committed:
Support WXQT in CMake

MaartenBent committed:
Do not use variable-length array in WXQT wxRegion

This is not supported by MSVC compiler.

MaartenBent committed:
Build WXQT toolkit with AppVeyor

MaartenBent committed:
Remove wxConsoleEventLoop from WXQT

The console event loop in the wxBase library is GUI toolkit independent.

MaartenBent committed:
Check if AppVeyor should abort before cloning the repo

MaartenBent committed:
Remove non-existing WXQT headers from build system

MaartenBent committed:
Fix several build errors in WXQT

Move declaration of wxOwnerDrawnBase::ms_defaultMargin to correct file.
Do not include headers when wxUSE_UIACTIONSIMULATOR is disabled.
Add guards for wxUSE_DRAG_AND_DROP.
Use a different wxFont constructor in printing sample, which is also available in WXQT.

MaartenBent committed:
Fix building stc library with WXQT

When using MSVC compiler, wxUSE_GRAPHICS_DIRECT2D is automatically enabled in setup.h.
But it is only available with WXMSW toolkit, not WXGTK or WXQT.

MaartenBent committed:
Fix missing gl functions in WXQT

vadz commented on src/stc/PlatWX.cpp:
```diff
@@ -44,7 +44,7 @@
 #include "wx/stc/private.h"
 
-#if wxUSE_GRAPHICS_DIRECT2D
+#if defined(__WXMSW__) && wxUSE_GRAPHICS_DIRECT2D
```
Could be worth introducing a special symbol for this (just in this file, not `setup.h`) to avoid repeating the same condition, e.g. `HAVE_DIRECTWRITE_TECHNOLOGY`?

vadz commented on tests/misc/safearrayconverttest.cpp:
```diff
@@ -12,7 +12,7 @@
 #endif
 
-#ifdef __WINDOWS__
+#if defined(__WINDOWS__) && !defined(__WXQT__)
```
Hmm, I wonder if wxQt shouldn't include the OLE stuff too? If this test compiles with wxGTK, it means that even it includes it under Windows, so why not Qt too?

vadz reviewed (commented):
Looks good to me, thanks!

Is this ready to merge now (after reverting Travis config removal)?

MaartenBent committed:
Fix missing wxOwnerDrawn functions in WXQT wxMenuItem

MaartenBent committed:
Include some Windows specific files in WXQT toolkit

Similar as with WXGTK toolkit on Windows.

MaartenBent committed:
Remove remnants of OS/2 from build system

MaartenBent commented:
I added some missing functions so all samples build successfully, with both CMake and configure. And also improved the WXGTK build on Windows with CMake.

I'll cleanup and squash some commits together later, and let you know when it is ready to merge.

MaartenBent committed:
Improve WXGTK build on Windows with CMake

MaartenBent commented:
You can merge if there are no objections.

vadz commented:
Great, thanks, just in time for #955. Will do in a moment.

Thanks again!
